src/pykeen/metrics/ranking.py

```diff
@@ -749,25 +852,90 @@ class GeometricMeanRank(RankBasedMetric):
     value_range = ValueRange(lower=1, lower_inclusive=True, upper=math.inf)
     increasing = False
     synonyms: ClassVar[Collection[str]] = ("gmr",)
+    supports_weights = True
+    closed_expectation: ClassVar[bool] = True
+    closed_variance: ClassVar[bool] = True
 
-    def __call__(self, ranks: np.ndarray, num_candidates: Optional[np.ndarray] = None) -> float:  # noqa: D102
-        return stats.gmean(ranks).item()
+    def __call__(
+        self, ranks: np.ndarray, num_candidates: Optional[np.ndarray] = None, weights: Optional[np.ndarray] = None
+    ) -> float:  # noqa: D102
+        return stats.gmean(ranks, weights=weights).item()
 
     def expected_value(
         self,
         num_candidates: np.ndarray,
         num_samples: Optional[int] = None,
+        weights: Optional[np.ndarray] = None,
+        **kwargs,
+    ) -> float:  # noqa: D102
+        is_log, individual = self._individual_expectation(num_candidates=num_candidates, weights=weights)
+        return stable_product(individual, is_log=is_log).item()
+
+    def variance(
+        self,
+        num_candidates: np.ndarray,
+        num_samples: Optional[int] = None,
+        weights: Optional[np.ndarray] = None,
         **kwargs,
     ) -> float:  # noqa: D102
+        # V (prod x_i) = prod (V[x_i] - E[x_i]^2) - prod(E[x_i])^2
+        is_log, individual_expectation = self._individual_expectation(num_candidates=num_candidates, weights=weights)
+        if is_log:
+            individual_expectation = np.exp(individual_expectation)
+        individual_variance = self._individual_variance(
+            num_candidates=num_candidates, weights=weights, individual_expectation=individual_expectation
+        )
+        return (
+            stable_product(individual_variance + individual_expectation**2)
+            - stable_product(individual_expectation) ** 2
+        )
+
+    @staticmethod
+    def _individual_variance(
+        num_candidates: np.ndarray, weights: np.ndarray, individual_expectation: np.ndarray
+    ) -> np.ndarray:
+        # use V[x] = E[x^2] - E[x]^2
+        x2 = (
+            np.exp(GeometricMeanRank._log_individual_expectation_no_weight(num_candidates=num_candidates, factor=2.0))
```

Comment:
make class method to refer to this

Reply:
You mean for 
`np.exp(GeometricMeanRank._log_individual_expectation_no_weight(num_candidates=num_candidates, factor=2.0))` ?

It is used only once, and already quite nested - I am not sure whether it would improve readability.